ipam/pool_test.go

```diff
@@ -0,0 +1,991 @@
+				It("Should return the pool", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id:       poolId,
+						RefCount: 0,
+					}
+					ap, err := as.requestPool(poolId, "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal(poolId))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When pool is in use", func() {
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id:       poolId,
+						RefCount: 1,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When pool is ipv4 and ipv6 is wanted", func() {
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id:     poolId,
+						IsIPv6: false,
+					}
+					ap, err := as.requestPool("", "", nil, true)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When IfName is not match", func() {
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					ifName := "eth0"
+					as.Pools[poolId] = &addressPool{
+						Id:     poolId,
+						IfName: ifName,
+					}
+					options := map[string]string{}
+					options[OptInterfaceName] = "en0"
+					ap, err := as.requestPool("", "", options, false)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When as has one pool available", func() {
+				It("Should return the pool", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id: poolId,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal(poolId))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When as has pools with different priority", func() {
+				It("Should return the pool with the highest priority", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["pool1"] = &addressPool{
+						Id:       "pool1",
+						Priority: 1,
+					}
+					as.Pools["pool2"] = &addressPool{
+						Id:       "pool2",
+						Priority: 2,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("pool2"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When as has pools with different addresses", func() {
```

Comment:
What about using the struct name `addressSpace` or `addressSpace object` instead of the variable name in a context, so that we can understand from the plain text context better? we can discuss here.
```suggestion
			Context("When a addressSpace has pools with different number of addresses", func() {
```

Reply:
If you agree, please also edit all other affected cases.

Reply:
agree, I also changed the ar to addressRecord